README.md
brainless
=========

#### Interpreter for the Brainfuck esoteric language ####

[![npm][npm-image]][npm-url]
[![david][david-image]][david-url]

[Wikipedia page](http://en.wikipedia.org/wiki/Brainfuck).

The following features are supported:

- Almost unlimited number of cells. The ECMA specification for the maximum array length is an unsigned 32-bit
integer, `2^32-1 = 4,294,967,295 cells`.  
  Negative pointer values are also valid, but they are stored in a separate tape. This means that the total tape length is `2*(2^32-1) = 8,589,934,590 cells`.  
  The state of the program when the pointer is outside these valid ranges is undefined.
- Each cell stores an IEEE 754 double-precision integer (64 bits), so negative numbers and UTF8 multibyte characters are allowed.
- It recognizes the usual tokens, but adds the de-facto `#` for debugging purposes. Prints a json object to the stdout, something similar to this: `{"pointer":-1,"tape":{"p":[97,98,99,0],"n":[null,0]}}`.
- It doesn't recognize the token `!`. There's a cli option to enter an input string.

Tokens:

<table>
  <tr>
    <td>&gt;</td>
    <td>Increments the pointer.</td>
  </tr>
  <tr>
    <td>&lt;</td>
    <td>Decrements the pointer.</td>
  </tr>
  <tr>
    <td>+</td>
    <td>Increments the cell value where points the pointer.</td>
  </tr>
  <tr>
    <td>-</td>
    <td>Decrements the cell value where points the pointer.</td>
  </tr>
  <tr>
    <td>.</td>
    <td>Outputs to the stdout the UTF8 character stored in the cell where points the pointer.</td>
  </tr>
  <tr>
    <td>,</td>
    <td>Reads a UTF8 character and stores it in the cell where points the pointer.</td>
  </tr>
  <tr>
    <td>[</td>
    <td>If the value in the cell where points the pointer is 0, it jumps to the ] token, otherwise it executes the code inside [ and ].</td>
  </tr>
  <tr>
    <td>]</td>
    <td>If the value in the cell where points the pointer is not 0, it jumps back again to the [ token, otherwise it continues with the program execution.</td>
  </tr>
  <tr>
    <td>#</td>
    <td>Outputs the pointer and the state of the tape. For debugging purposes.</td>
  </tr>
</table>

#### Usage ####

If the program needs to read an input value but no string is provided, it prompts a cli message asking for a character:

```
$ brainfuck examples/rot13
> a
n
> b
o
> c
p
```

Send a SIGINT signal (ctrl+c) to kill the process.

__CLI__

```
$ brainfuck -h
Usage: brainfuck [options] <input_file>

Interpreter for the Brainfuck esoteric language

  -i, --input=STRING          STRING to read during the program execution
  -h, --help                  Display this help message and exit
  -v, --version               Output version information and exit

Report bugs to <gagle@outlook.com>.
```

For example:

```
$ brainfuck examples/reverse -i abc
cba
```

Note: If the file doesn't exist and doesn't have an extension, it tries to read the same file with the extension `.bf`.

__Programmatically__

___module_(code[, input]) : undefined__  
Executes the given code.

```javascript
var bf = require('brainless');
bf(',[>,]<[.<]', 'abc');
// cba
```

[npm-image]: http://img.shields.io/npm/v/bole-mongodb.svg?style=flat
[npm-url]: https://npmjs.org/package/bole-mongodb
[david-image]: https://img.shields.io/david/gagle/node-brainless.svg?style=flat
[david-url]: https://david-dm.org/gagle/node-brainless

CHANGES
v0.1.0 (16 Jan 2015)
  Clean up and bump version.

v0.0.4 (25 Nov 2013)
  Minor changes.

v0.0.4 (25 Nov 2013)
  Minor changes.

v0.0.3 (29 Oct 2013)
  Minor improvements.

v0.0.2 (25 Oct 2013)
  Added negative pointers.
  Minor improvements.

v0.0.1 (25 Oct 2013)
  First release.

lib/index.js
'use strict';

/*
Almost unlimited number of cells: ECMA, array length: unsigned 32-bit
integer, 2^32-1 = 4,294,967,295.

Negative pointer values are also valid, but they are stored in a separate
tape. This means that the total tape length is 2*2^32-1 = 8,589,934,590.

The state of the program when the pointer is outside these valid ranges is
undefined.

Each cell stores an IEEE 754 double-precision integer (64 bits), so negative
numbers and utf8 multibyte characters are allowed.
*/

var readline = require('readline');

var syntaxError = function (message, line, column) {
  console.error('SyntaxError: %s line %d, column %d', message, line, column);
  process.exit(1);
};

module.exports = function (code, input) {
  var stdin = typeof input !== 'string';
  var meta = {
    pointer: 0,
    tape: {
      p: [0],
      n: []
    }
  };
  var stack = [];
  var line = 1;
  var column = 1;
  var c;
  var resumeIndex;
  var skip;
  var stdout;

  var prompt = function () {
    // Hack to wake up the stdout stream, otherwise it doesn't print messages
    if (prompt._called) {
      console.log();
    } else {
      prompt._called = true;
    }
    var rl = readline.createInterface({
      input: process.stdin,
      output: process.stdout
    });
    rl.setPrompt('> ', 2);
    rl.on('line', function (line) {
      if (!line) {
        // Empty line, prompt again
        rl.setPrompt('> (again) ');
        rl.prompt();
        return;
      }
      meta.tape.p[meta.pointer] = line.charCodeAt(0);
      this.close();
    });
    rl.on('close', function () {
      run(resumeIndex);
    });
    rl.on('SIGINT', function () {
      process.stdout.write('\n');
      process.exit();
    });
    rl.prompt();
  };

  var run = function (start) {
    var t;
    var p;

    for (var i = start; i < code.length; i++, column++) {
      c = code.charCodeAt(i);

      // Skip the tokens until ], this only applies when the loop is never
      // executed
      if (skip) {
        if (c === 91) {
          // [
          skip.depth++;
        } else if (c === 93) {
          // ]
          if (!skip.depth) {
            skip = null;
          } else {
            skip.depth--;
          }
        } else if (c === 10) {
          // \n
          line++;
          column = 0;
        }
        continue;
      }

      if (c === 43) {
        // +
        if (meta.pointer >= 0) {
          meta.tape.p[meta.pointer]++;
        } else {
          meta.tape.n[-meta.pointer]++;
        }
      } else if (c === 45) {
        // -
        if (meta.pointer >= 0) {
          meta.tape.p[meta.pointer]--;
        } else {
          meta.tape.n[-meta.pointer]--;
        }
      } else if (c === 62) {
        // >
        // Set to 0 if it's the first access
        if (++meta.pointer >= 0) {
          if (meta.tape.p[meta.pointer] === undefined) {
            meta.tape.p[meta.pointer] = 0;
          }
        } else {
          if (meta.tape.n[-meta.pointer] === undefined) {
            meta.tape.n[-meta.pointer] = 0;
          }
        }
      } else if (c === 60) {
        // <
        // Set to 0 if it's the first access
        if (--meta.pointer < 0 && meta.tape.n[-meta.pointer] === undefined) {
          meta.tape.n[-meta.pointer] = 0;
        }
      } else if (c === 91) {
        // [
        if (meta.pointer >= 0) {
          t = meta.tape.p;
          p = meta.pointer;
        } else {
          t = meta.tape.n;
          p = -meta.pointer;
        }

        if (t[p]) {
          stack.push({
            start: i,
            line: line,
            column: column
          });
        } else {
          // Skip to the closing ]
          skip = {
            line: line,
            column: i + 1,
            depth: 0
          };
        }
      } else if (c === 93) {
        // ]
        var el = stack.pop();
        if (el === undefined) {
          // Missing openning [
          syntaxError('Missing openning \'[\' for matching \']\' at', line,
              column);
        } else {
          if (meta.pointer >= 0) {
            t = meta.tape.p;
            p = meta.pointer;
          } else {
            t = meta.tape.n;
            p = -meta.pointer;
          }

          if (t[p]) {
            // Jump back
            stack.push(el);
            i = el.start;
            line = el.line;
            column = el.column;
          }
        }
      } else if (c === 46) {
        // .
        stdout = true;
        if (meta.pointer >= 0) {
          t = meta.tape.p;
          p = meta.pointer;
        } else {
          t = meta.tape.n;
          p = -meta.pointer;
        }
        process.stdout.write(String.fromCharCode(t[p]));
      } else if (c === 44) {
        // ,
        if (stdin) {
          resumeIndex = i + 1;
          prompt();
          return;
        } else if (input[0]) {
          if (meta.pointer >= 0) {
            t = meta.tape.p;
            p = meta.pointer;
          } else {
            t = meta.tape.n;
            p = -meta.pointer;
          }
          t[p] = input[0].charCodeAt(0);
          input = input.slice(1);
        }
        // If there's no more input data to read, this token is no-op
      } else if (c === 35) {
        // #
        var msg = '';
        if (stdout) msg += '\n';
        stdout = true;
        process.stdout.write(msg + JSON.stringify(meta) + '\n');
      } else if (c === 10) {
        // \n
        line++;
        column = 0;
      }
    }

    // Check missing ]
    if (skip) {
      syntaxError('Missing closing \']\' for matching \'[\' at', skip.line,
          skip.column);
    } else if (stack.length) {
      syntaxError('Missing closing \']\' for matching \'[\' at', stack[0].line,
          column);
    }
  };

  run(0);
};